### PR TITLE
[JENKINS-34545] - Support case insensitive user IDs in roles

### PR DESCRIPTION
**Pull request by seanmceligot**
This should fix JENKINS-34545 and related issues. It allows case insensitive matching to user/group values for Global and Item roles in Assign Roles.

### CODE REVIEW

**Comment by seanmceligot** [comment]
OK, I'll look into that.

**Comment by seanmceligot** [comment]
Here's a screenshot of what I have now. I was thinking Assign Roles screen is the most relevant place to put this but let me know what you think. Current it doesn't reload map without restarting Jenkins. I'm looking into that.

roles

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/Role.java [hunk 1]
```diff
@@ -223,4 +223,11 @@ public boolean equals(Object obj) {
         }
         return true;
     }
+  public String toString() {
+        StringBuilder bldr = new StringBuilder("Role:");
+        bldr.append(name);
+        bldr.append(':');
+        bldr.append(pattern);
+        return bldr.toString();
```
Comment (KostyaSha) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/Role.java]: with String.format() you can do it in one line

**Comment by oleg-nenashev** [comment]
Maybe it makes sense to move the setting to the common Authorization Strategy configuration screen (`config.jelly` in https://github.com/jenkinsci/role-strategy-plugin/tree/master/src/main/resources/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy) and making the setter a `@DataBoundSetter`. Otherwise this option will immediately become unmanageable from Configuration as Code.

No strong opinion tho, it is a wider problem than this PR

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java [hunk 2]
```diff
@@ -78,8 +78,10 @@
           .expireAfterWrite(Settings.USER_DETAILS_CACHE_EXPIRATION_TIME_SEC, TimeUnit.SECONDS)
           .build();
 
+  private final transient boolean caseInsensitiveUser;
```
Comment (oleg-nenashev) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: I am not sure what is the intention here. With the current logic the flag will be always false after the restart, and adding new roles from API won't behave as expected

**Comment by daniel-beck** [comment]
Why not integrate with case sensitivity of the chosen security realm? Or what am I missing?

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java (hunk 2 above)
Comment (seanmceligot) [reply src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: The real configuration variable is in RoleBasedAuthorizationStrategy. That one is currently marshed/unmarshed explicitly in that RoleBasedAuthorizationStrategy.ConverterImpl. I just followed what was already there. Does the ConvererImpl entirely need to be converted or do you want the new variable managed elsewhere? 

This RoleMap.caseInsensitiveUser is a copy needed to create the TreeSet  when RoleMap.addRole is called. I made it transient because I didn't want it stored in config.xml and final to show that it was not the config.xml variable. It loads correctly at startup from config.xml, but I need to recreate all the TreeSets when it's changed realtime. 

config.xml
  ```<authorizationStrategy 
  class="com.michelin.cio.hudson.plugins.rolestrategy.RoleBasedAuthorizationStrategy">
      <caseInsensitiveUser>true</caseInsensitiveUser>
```

new variable placement on Configure Global Security page.

caseinsensitveconfiguresecurity

**Comment by ingwarsw** [comment]
@oleg-nenashev Any plans to integrate this..
Its really hard to explain users to login using always the same case..

**Comment by oleg-nenashev** [comment]
No short-term plans. If somebody wants to contribute, happy to share my
vision

**Comment by seanmceligot** [comment]
Just an update. My original fix was two lines lines of code.  

https://github.com/jenkinsci/role-strategy-plugin/compare/master...seanmceligot:case-insensitive-2.9.0
  
I started to add an configuration option turn it on and off as requested for legacy users, but my change wasn't compatible with configuration as code. I'd need to delve into the CASC framework if I wanted to continue.

**Comment by kiranraj365** [comment]
@seanmceligot  can we expect any progress on this issue in future.

**Comment by seanmceligot** [comment]
> @seanmceligot can we expect any progress on this issue in future.

Now should be a good time since configuration as code was merged. Let me look at adding the configuration parameter again and make sure it's CASC compatible.

**Comment by kiran1abc** [comment]
@seanmceligot thanks that helps us a lot.

**Comment by seanmceligot** [comment]
screenshot

**Comment by oleg-nenashev** [comment]
Thanks @seanmceligot ! It would be really great to get it over the line

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java [hunk 3]
```diff
@@ -141,7 +158,7 @@ private boolean hasPermission(String sid, Permission permission, RoleType roleTy
     new RoleWalker() {
       public void perform(Role current) {
         if (current.hasAnyPermission(permissions)) {
-          if (grantedRoles.get(current).contains(sid)) {
+          if (hasSid( grantedRoles.get(current), userIdStrategy, sid))  {
```
Comment (AbhyudayaSharma) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: `contains()` in a set is a relatively fast operation. Now, you have to traverse the entire set to see if the role actually contains the sid.

The benchmarks suggest the same:
image
(The results were compared to the last successful build of master)

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java [hunk 4]
```diff
@@ -127,12 +141,15 @@ public RoleMap(@Nonnull SortedMap<Role,Set<String>> grantedRoles) {
   public static void invalidateDangerousPermissions() {
     PermissionHelper.DANGEROUS_PERMISSIONS.forEach(implyingPermissionCache::remove);
   }
-
+  private boolean hasSid(Set<String> roleAssignements, IdStrategy userIdStrategy, String sid) {
```
Comment (AbhyudayaSharma) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: Typo here: `roleAssignements` → `roleAssignments`

File: src/test/resources/org/jenkinsci/plugins/rolestrategy/Configuration-as-Code.yml [hunk 5]
```diff
@@ -1,6 +1,7 @@
 jenkins:
   authorizationStrategy:
     roleBased:
+      userIdStrategyChoice: DEFAULT
```
Comment (AbhyudayaSharma) [inline src/test/resources/org/jenkinsci/plugins/rolestrategy/Configuration-as-Code.yml]: Shouldn't we have a test case where the `userIdStrategyChoice` is not present?

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java (hunk 3 above)
Comment (seanmceligot) [reply src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: Yes, I can't use Set because it's optimized at creation time time but IDStrategy as a configuration can be changed anytime.  I suppose I could use a case insensitive Set and then verify it using the correct IDStrategy at lookup time. Or, rebuild the Set every time the configuration changes, but that's a but trickier to get right.

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java (hunk 4 above)
Comment (seanmceligot) [reply src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: fixed. Thank you. https://github.com/seanmceligot/role-strategy-plugin/commit/3be3e3bc8a41b3b713f07d13c03ef2cbb99cec31

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java [hunk 6]
```diff
@@ -116,13 +142,49 @@ public RoleBasedAuthorizationStrategy(Map<String, RoleMap> grantedRoles) {
...
+  }
+  @DataBoundSetter
+  public void setUserIdStrategyChoice(UserIdStrategyChoice userIdStrategyChoice) {
+      LOGGER.info("setUserIdStrategyChoice: "+ userIdStrategyChoice);
+      this.userIdStrategyChoice = userIdStrategyChoice;
+      globalRoles.invalidateCache();
+      agentRoles.invalidateCache();
+      itemRoles.invalidateCache();
+  }
+  public void setUserIdStrategyChoiceStr(String userIdStrategyChoice) {
```
Comment (oleg-nenashev) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java]: I suggest not making such public method. 

```suggestion
  @Restricted(NoExternalUse.class)
  public void setUserIdStrategyChoiceStr(String userIdStrategyChoice) {
```

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java [hunk 7]
```diff
@@ -116,13 +142,49 @@ public RoleBasedAuthorizationStrategy(Map<String, RoleMap> grantedRoles) {
...
+      LOGGER.info("setUserIdStrategyChoice"+ userIdStrategyChoice);
+      setUserIdStrategyChoice( UserIdStrategyChoice.valueOf(userIdStrategyChoice) );
```
Comment (oleg-nenashev) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java]: What if a non-existent value is passed. It will fail here, right?

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java [hunk 8]
```diff
@@ -680,6 +757,7 @@ public void doRolesSubmit(StaplerRequest req, StaplerResponse rsp) throws Unsupp
 
         req.setCharacterEncoding("UTF-8");
       JSONObject json = req.getSubmittedForm();
+      LOGGER.info("doRolesSubmit: "+json.toString());
```
Comment (oleg-nenashev) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java]: Unrelated. If you need it, please use `FINEST` level or lower

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java [hunk 9]
```diff
@@ -696,6 +774,7 @@ public void doAssignSubmit(StaplerRequest req, StaplerResponse rsp) throws Unsup
 
         req.setCharacterEncoding("UTF-8");
       JSONObject json = req.getSubmittedForm();
+      LOGGER.info("doAssignSubmit: "+json.toString());
```
Comment (oleg-nenashev) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java]: same

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java (hunk 3 above)
Comment (oleg-nenashev) [reply src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: Rebuilding set would be a good approach IMO.
I would be fine if the entire `RoleStrategy` object gets rebuilt on a configuration change

File: src/main/java/org/jenkinsci/plugins/rolestrategy/casc/RoleBasedAuthorizationStrategyConfigurator.java [hunk 10]
```diff
@@ -33,6 +36,8 @@
 @Restricted({NoExternalUse.class})
 public class RoleBasedAuthorizationStrategyConfigurator extends BaseConfigurator<RoleBasedAuthorizationStrategy> {
 
+	  private static final Logger LOGGER = Logger.getLogger(RoleBasedAuthorizationStrategyConfigurator.class.getName());
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/rolestrategy/casc/RoleBasedAuthorizationStrategyConfigurator.java]: tabs vs. spaces

```suggestion
   private static final Logger LOGGER = Logger.getLogger(RoleBasedAuthorizationStrategyConfigurator.class.getName());  
```

File: src/test/java/org/jenkinsci/plugins/rolestrategy/RoleStrategyTest.java [hunk 11]
```diff
@@ -7,8 +7,7 @@
 import hudson.model.FreeStyleProject;
 import hudson.model.Item;
 import hudson.model.User;
-import hudson.security.AuthorizationStrategy;
-import io.jenkins.plugins.casc.ConfigurationContext;
+import hudson.security.AuthorizationStrategy; import io.jenkins.plugins.casc.ConfigurationContext;
```
Comment (oleg-nenashev) [inline src/test/java/org/jenkinsci/plugins/rolestrategy/RoleStrategyTest.java]:
```suggestion
import hudson.security.AuthorizationStrategy;
import io.jenkins.plugins.casc.ConfigurationContext;
```

File: src/main/resources/com/michelin/cio/hudson/plugins/rolestrategy/RoleStrategyConfig/assign-roles.jelly [hunk 12]
```diff
@@ -56,7 +56,7 @@
               <td class="left-most">${title}</td>
               <j:forEach var="r" items="${it.strategy.getGrantedRoles(attrs.type)}">
                 <td width="*">
-                  <f:checkbox name="[${r.key.name}]" checked="${r.value.contains(attrs.sid)}"/>
+                    <f:checkbox name="[${r.key.name}]" checked="${r.value.contains(attrs.sid)}"/>
```
Comment (oleg-nenashev) [inline src/main/resources/com/michelin/cio/hudson/plugins/rolestrategy/RoleStrategyConfig/assign-roles.jelly]: Revert it 

```suggestion
                  <f:checkbox name="[${r.key.name}]" checked="${r.value.contains(attrs.sid)}"/>
```

File: src/test/resources/org/jenkinsci/plugins/rolestrategy/Configuration-as-Code.yml (hunk 5 above)
Comment (oleg-nenashev) [reply src/test/resources/org/jenkinsci/plugins/rolestrategy/Configuration-as-Code.yml]: We should. Also, it is odd that all tests require "DEFAULT" and none really tests values set

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java [hunk 13]
```diff
@@ -127,12 +141,15 @@ public RoleMap(@Nonnull SortedMap<Role,Set<String>> grantedRoles) {
   public static void invalidateDangerousPermissions() {
     PermissionHelper.DANGEROUS_PERMISSIONS.forEach(implyingPermissionCache::remove);
   }
-
+  private boolean hasSid(Set<String> roleAssignments, IdStrategy userIdStrategy, String sid) {
+    return roleAssignments.stream().anyMatch( rolesid -> userIdStrategy.compare(rolesid, sid) == 0 );
```
Comment (AbhyudayaSharma) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: I was going through how Matrix Authorization plugin handles it. Looks to me that this check is incomplete. Since these sids can be group names too and there exists `SecurityRealm#getGroupIdStrategy()`, you will need to take care of that.

https://github.com/jenkinsci/matrix-auth-plugin/blob/8c7b7bcec67a03ab73498486fc2029b2b886541d/src/main/java/org/jenkinsci/plugins/matrixauth/AuthorizationContainer.java#L152

Comment (seanmceligot) [reply src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: I will look that at. Good catch.

**Comment by zhaojinggang** [comment]
> This should fix JENKINS-34545 and related issues. It allows case insensitive matching to user/group values for Global and Item roles in Assign Roles.

Just check if there's any plan to release this feature?
